### DISCUSSION TITLE
Heap CONFIGs

### DISCUSSION THREAD

**Discussion by ExtremeGTX** [zephyrproject-rtos/zephyr#53842] [Q&A]
Hi
I want to understand why Zephyr doesn't have a single shared heap rather (as far as i could discover) there are:
- CONFIG_MBEDTLS_HEAP_SIZE: used by mbedTLS
- CONFIG_HEAP_MEM_POOL_SIZE: used by k_malloc
- CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE: used for malloc from libc
is it possible to at least make `malloc` and `k_malloc` use the same heap ?
is it designed this way for some security concerns ?
Thanks in advance :slightly_smiling_face:

**Comment by yashi** [comment]
To avoid runtime allocation and to provide data protection, I suppose.

Linux "kernel" does have one memory pool but applications uses it's own heap, which is in turn allocated from the kernel at _runtime_.  Because on a typical Linux system applications comes and goes.  It is the kernel's job to manage all memory in the system for them.

Zephyr, instead, provides compile time heap allocation by `K_HEAP_DEFINE`.  Because for embedded systems, it usually runs with fixed number of threads, and does't need runtime heap allocation.

Runtime allocation means fragmentation.  Systems which Zephyr targets tend to have smaller amount of memory. And we want to use it as efficient as possible.  Because a larger memory means higher costs in embedded systems.

**Reply by ExtremeGTX** [reply]
Thanks yashi.
I know in embedded systems we try to avoid dynamic allocation but sometimes it is inevitable.
Still my question is why multiple Heap spaces ? isn't that increases memory fragmentation and doesn't improve overall utilization of the heap space?
What are the disadvantages of merging all of them in a single heap managed by a single allocator?

**Reply by lucasgcbkhomp** [reply]
My understanding of it is that, with multiple heaps, you can better manage how much you'll be allocating at a time from each reserved heap space, avoiding fragmentation. If all Heaps were combined into one, you'd risk different allocation sizes being performed on the same address space, which eventually leads to fragmentation.

**Reply by yashi** [reply]
"Fragmentation" has two basic types, internal and external.  Multiple heaps creates the internal fragmentation and a single heap creates external fragmentation.  It's not like one is superior than the other.  They have different properties.

With a single heap, as @lucasgcbkhomp said, you need more complex heap management code to _efficiently_ satisfy different allocation sizes from variety of subsystems. That means larger code size and more runtime cost.

Zephyr, I suppose, has chosen to have multiple heaps to cut down this code complexity and selected the simplicity.  And with that we also gained [memory protection with separation of memory locations](https://docs.zephyrproject.org/latest/kernel/usermode/memory_domain.html#memory-domains) for each domain / subsystem.
